crypto/bn/bn_dsa.c

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <openssl/opensslconf.h>
+#ifdef FIPS_MODE
+NON_EMPTY_TRANSLATION_UNIT
+#else
+# include <stdio.h>
+# include <string.h>
+# include <openssl/evp.h>
+# include <openssl/rand_drbg.h>
+# include "internal/cryptlib.h"
+# include "internal/bn_int.h"
+# include "internal/thread_once.h"
+# include "../rand/rand_lcl.h"
+
+typedef struct dsa_drbg_ctx_st
+{
+    const unsigned char *x;
+    const unsigned char *h;
+    size_t rlen;
+} DSA_DRBG_CTX;
+
+static int app_data_index = -1;
+static CRYPTO_ONCE deterministic_dsa_init = CRYPTO_ONCE_STATIC_INIT;
+
+DEFINE_RUN_ONCE_STATIC(do_deterministic_dsa_init)
```

Comment:
Drive by comment: If this is going to be used within the FIPS module, then "RUN_ONCE" is not a suitable mechanism (probably we ought to prevent its use in the module completely). In fact, now that we have OPENSSL_CTX I would suggest we avoid RUN_ONCE entirely (whether or not the code is being imported into the FIPS module).

See the following for an example of storing "global" data in an OPENSSL_CTX rather than via RUN_ONCE:

https://github.com/openssl/openssl/blob/7a2027240e1d01f7f5b209998d1de36af221b34b/crypto/rand/drbg_lib.c#L213-L241

https://github.com/openssl/openssl/blob/7a2027240e1d01f7f5b209998d1de36af221b34b/crypto/rand/drbg_lib.c#L250-L258

Reply:
I created PR #9308 to prevent the accidental use of RUN_ONCE inside the FIPS module.